Add `obj-det` pipeline support for `LayoutLMV2`

mishig25 opened the pull request:
# What does this PR do?

Note: I'm using terms `document-understanding` & `layout-detection` interchangeably and imo,  term `layout-detection` sounds more accurate.

As we have discussed in huggingface/hub-docs#21, reusing `object-detection` pipeline for `layout-detection` architectures (specifically, `LayoutLMv2ForTokenClassification).

An important detail in reusing so is that:
* `LayoutLMv2ForTokenClassification` needs `LayoutLMv2Processor` to preprocess input image
* Since processor is just a combination of (tokenizer + feature_extractor), `ObjectDetectionPipeline.preprocess` does exactly what `LayoutLMv2Processor` does when the selected model has architecture `...ForTokenClassification`

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes # (issue)


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?


## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- albert, bert, xlm: @LysandreJik
- blenderbot, bart, marian, pegasus, encoderdecoder,  t5: @patrickvonplaten, @patil-suraj
- longformer, reformer, transfoxl, xlnet: @patrickvonplaten
- fsmt: @stas00
- funnel: @sgugger
- gpt2: @patrickvonplaten, @LysandreJik
- rag: @patrickvonplaten, @lhoestq
- tensorflow: @LysandreJik

Library:

- benchmarks: @patrickvonplaten
- deepspeed: @stas00
- ray/raytune: @richardliaw, @amogkam
- text generation: @patrickvonplaten
- tokenizers: @n1t0, @LysandreJik
- trainer: @sgugger
- pipelines: @LysandreJik

Documentation: @sgugger

HF projects:

- datasets: [different repo](https://github.com/huggingface/datasets)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Examples:

- maintained examples (not research project or legacy): @sgugger, @patil-suraj
- research_projects/bert-loses-patience: @JetRunner
- research_projects/distillation: @VictorSanh

 -->
@Narsil @NielsRogge

mishig25 committed:
Add obj-det pipe support for LayoutLMV2

mishig25 committed:
Chore

NielsRogge commented on src/transformers/models/layoutlmv2/feature_extraction_layoutlmv2.py:
```diff
@@ -22,11 +22,15 @@
 from ...utils import logging
 
 
+if is_torch_available():
+    import torch
+    from torch import nn
+
```
Can you still import `LayoutLMv2FeatureExtractor` when you don't have torch installed?

NielsRogge replied:
Can you check this @mishig25?

mishig25 replied:
Not sure if this answers your question. Added import torch lines because the newly added `post_process` method depends on torch
https://github.com/huggingface/transformers/blob/01509bf3a6851fafaa9ad0ca9deecaa05bdf6ea4/src/transformers/models/layoutlmv2/feature_extraction_layoutlmv2.py#L239-L244

Should I move the import torch statement inside `post_process` method ?

NielsRogge replied:
cc'ing @LysandreJik here to check what's the best option

LysandreJik replied:
I would keep the import torch statement at the top level. Moving it inside the `post_process` function means it might crash once a program is well into progress, which can be painful. I'd rather it fail early on.

NielsRogge commented on src/transformers/models/layoutlmv2/feature_extraction_layoutlmv2.py:
```diff
@@ -220,3 +233,54 @@ def __call__(
+
+    def post_process(self, outputs, target_sizes, offset_mapping, bbox):
+        """
+        Converts the output of :class:`~transformers.DetrForObjectDetection` into the format expected by the COCO api.
```
```suggestion
        Converts the output of :class:`~transformers.LayoutLMv2ForTokenClassification` into the format expected by the COCO api.
```

NielsRogge commented on src/transformers/pipelines/object_detection.py:
```diff
@@ -108,22 +108,42 @@ def preprocess(self, image):
         image = self.load_image(image)
         target_size = torch.IntTensor([[image.height, image.width]])
         inputs = self.feature_extractor(images=[image], return_tensors="pt")
+        if self._is_layout_detection():
+            encoded_inputs = self.tokenizer(
+                text=inputs["words"],
+                boxes=inputs["boxes"],
+                truncation=True,
+                return_offsets_mapping=True,
+                return_tensors="pt",
+            )
+            encoded_inputs["image"] = inputs.pop("pixel_values")
+            encoded_inputs["target_size"] = target_size
+            return encoded_inputs
```
I see you're using the feature extractor and tokenizer separately here. Note that you can just use the processor instead:
```suggestion
        if self._is_layout_detection():
            encoded_inputs = self.processor(image, truncation=True, return_offsets_mapping=True, return_tensors="pt")
            return encoded_inputs
```

Narsil replied:
Because pipelines don't want to rely on the `processor`. `processor` is a concept that does not exist for most models, all the code exists to support loading `tokenizers` and `feature_extractor` automatically and magically, so using `processor` is really not worth the effort (especially since we're using a very small subset of what `processor` logic does)

NielsRogge replied:
Ok, however at some point I guess we will have an `AutoProcessor` API that will load the appropriate preprocessing thing (either a `feature_extractor`, a `tokenizer` or a `processor`). For now, using the `feature_extractor` and `tokenizer` separately is fine.

mishig25 replied:
Closing it as resolved for now

mishig25 committed:
Update src/transformers/models/layoutlmv2/feature_extraction_layoutlmv2.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

Narsil reviewed (commented):
Some suggestions to improve, but overall seems to work !

Cheers.

mishig25 commented on tests/test_pipelines_object_detection.py:
```diff
@@ -251,3 +254,64 @@ def test_threshold(self):
+        tokenizer = AutoTokenizer.from_pretrained(model_id)
+        object_detector = ObjectDetectionPipeline(
+            model=model, feature_extractor=feature_extractor, tokenizer=tokenizer
+        )
```
It works and was used in the test_method below 😊
https://github.com/huggingface/transformers/blob/fd7d47b301b37016212e6002098beefaf3863982/tests/test_pipelines_object_detection.py#L263-L268

Narsil replied:
Wow black magic does work ! :D

NielsRogge commented on src/transformers/models/auto/modeling_auto.py:
```diff
@@ -263,6 +263,7 @@
     [
         # Model for Object Detection mapping
         ("detr", "DetrForObjectDetection"),
+        ("layoutlmv2", "LayoutLMv2ForTokenClassification"),
```
Great point!

mishig25 replied:
Done

mishig25 commented on src/transformers/pipelines/object_detection.py:
```diff
@@ -108,22 +108,42 @@ def preprocess(self, image):
+
+        scores, labels, boxes = [raw_annotation[key] for key in ["scores", "labels", "boxes"]]
+        if not self._is_layout_detection():
+            keep = raw_annotation["scores"] > threshold
+            scores, labels, boxes = scores[keep], labels[keep], boxes[keep]
```
DetrFeatureExtractor.post_process & LayoutLmv2Extractor.post_process both produce outputs in COCO format (which is {score, labels, box} for every detected object)

ObjectDetectionPipelines.post_process filters DetrFeatureExtractor.post_process output (i.e. keep detections whose score > threshold). 
However, I'm turning off the filtering on outputs from LayoutLmv2Extractor.post_process (i.e. keep all detections)

Narsil replied:
Why ?

NielsRogge replied:
I wouldn't do that, the `threshold` parameter should be used. In the inference widget, you can then set threshold to 0 in order to keep all detections.

Narsil replied:
I would like to add, we should have sane defaults that work on every pipeline. If `0.9` is too high for that particular pipeline or something, we should look either:
- better parameters
- better defaults
- accepting defaults are not ok
we should strive to limit amount of `ifs` around parameters as it leads to unexpected results for users.

mishig25 replied:
0.9 was too high for `LayotLMv2ForTokenClassification` output; I'll test what should be a good `threhsold` parameter that can work both for Detr and LayoutLvm2

Narsil replied:
I see, if you can't find anything, maybe having a `top_k` parameters makes it easier to be pipeline agnostic.

mishig25 replied:
changed the default threshold to 0.5 for 2 reasons:
1. Detr softmaxed scores are almost [0,1] binary (i.e. the scores are either ~0 or ~1; so changing threshold from 0.9 to 0.5 doesn't affect post_process of Detr)
2. On the other hand, threshold of 0.5 keeps more of the predictions from LayoutLMV2ForTokenClass. 

And as Niels said, a user can always set threshold=0.0 to keep all predictions

mishig25 commented on src/transformers/models/layoutlmv2/feature_extraction_layoutlmv2.py:
```diff
@@ -47,6 +51,15 @@ def normalize_box(box, width, height):
 
 
+def unnormalize_box(box, height, width):
+    return [
```
@NielsRogge could you help on this one?

NielsRogge replied:
The 1000 comes from the fact that all coordinates are normalized to be in the range 0-1000, before being passed to the model. Afterwards, we "unnormalize" them again to their original values.

mishig25 replied:
Added the comment

Narsil replied:
So basicall model representaiton is [1000, 1000] square vs [1, 1] square of internal representation, that is rescaled back to [height, width] at the end ?

NielsRogge commented on src/transformers/models/layoutlmv2/feature_extraction_layoutlmv2.py:
```diff
@@ -220,3 +233,54 @@ def __call__(
+                image size (before any data augmentation). For visualization, this should be the image size after data
+                augment, but before padding.
+            offset_mapping (:obj:`torch.Tensor` of shape :obj:`(batch_size, x, 2)`):
+                TTensor coming that is "offset_mapping" field of outputs from
```
```suggestion
                Tensor coming that is "offset_mapping" field of outputs from
```

NielsRogge commented on src/transformers/models/layoutlmv2/feature_extraction_layoutlmv2.py:
```diff
@@ -220,3 +233,54 @@ def __call__(
+                Tensor containing the size (h, w) of each image of the batch. For evaluation, this must be the original
+                image size (before any data augmentation). For visualization, this should be the image size after data
+                augment, but before padding.
+            offset_mapping (:obj:`torch.Tensor` of shape :obj:`(batch_size, x, 2)`):
+                TTensor coming that is "offset_mapping" field of outputs from
+                :class:`~transformer.LayoutLMv2TokenizerFast`.
```
```suggestion
            offset_mapping (:obj:`torch.Tensor` of shape :obj:`(batch_size, x, 2)`):
                Tensor coming from the "offset_mapping" field of the outputs of :class:`~transformer.LayoutLMv2TokenizerFast`.
```

NielsRogge commented on src/transformers/models/layoutlmv2/feature_extraction_layoutlmv2.py:
```diff
@@ -220,3 +233,54 @@ def __call__(
+        Only supports PyTorch.
+
+        Args:
+            outputs :obj:`Dict`):
```
```suggestion
            outputs (:obj:`Dict`):
```

NielsRogge commented on src/transformers/models/layoutlmv2/feature_extraction_layoutlmv2.py:
```diff
@@ -220,3 +233,54 @@ def __call__(
+            outputs :obj:`Dict`):
+                Raw outputs of the model.
+            target_sizes (:obj:`torch.Tensor` of shape :obj:`(batch_size, 2)`):
+                Tensor containing the size (h, w) of each image of the batch. For evaluation, this must be the original
+                image size (before any data augmentation). For visualization, this should be the image size after data
+                augment, but before padding.
```
```suggestion
                Tensor containing the original size (h, w) of each image of the batch. 
```
LayoutLMv2 doesn't require any data augmentation, so we can leave that out.

NielsRogge commented on src/transformers/models/layoutlmv2/feature_extraction_layoutlmv2.py:
```diff
@@ -220,3 +233,54 @@ def __call__(
+                TTensor coming that is "offset_mapping" field of outputs from
+                :class:`~transformer.LayoutLMv2TokenizerFast`.
+            bbox (:obj:`torch.Tensor` of shape :obj:`(batch_size, x, 4)`):
+                Tensor coming that is "bbox" field of outputs from :class:`~transformer.LayoutLMv2TokenizerFast`.
```
```suggestion
                Tensor coming from the "bbox" field of the outputs of :class:`~transformer.LayoutLMv2TokenizerFast`.
```

NielsRogge commented on src/transformers/models/layoutlmv2/feature_extraction_layoutlmv2.py:
```diff
@@ -220,3 +233,54 @@ def __call__(
+
+        results = []
+
+        for s, l, b, _is_subword, (height, width) in zip(scores, labels, boxes, is_subword, target_sizes):
+            # only keep non-subword predictions
+            l = [label for idx, label in enumerate(l) if not _is_subword[idx]]
+            s = [score for idx, score in enumerate(s) if not _is_subword[idx]]
+            b = [unnormalize_box(box, height, width) for idx, box in enumerate(b) if not _is_subword[idx]]
+            results.append(
+                {"scores": torch.FloatTensor(s), "labels": torch.IntTensor(l), "boxes": torch.FloatTensor(b)}
+            )
+        return results
```
```suggestion
        for s, l, b, is_subword_example, (height, width) in zip(scores, labels, boxes, is_subword, target_sizes):
            # only keep non-subword predictions
            l = [label for idx, label in enumerate(l) if not is_subword_example[idx]]
            s = [score for idx, score in enumerate(s) if not is_subword_example[idx]]
            b = [unnormalize_box(box, height, width) for idx, box in enumerate(b) if not is_subword_example[idx]]
            results.append(
                {"scores": torch.FloatTensor(s), "labels": torch.IntTensor(l), "boxes": torch.FloatTensor(b)}
            )
        return results
```
This is for each example in the batch, right? Then a better name might be `is_subword_example`.

mishig25 committed:
Add box normalization comments

mishig25 committed:
Update src/transformers/models/layoutlmv2/feature_extraction_layoutlmv2.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

mishig25 committed:
Update src/transformers/models/layoutlmv2/feature_extraction_layoutlmv2.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

mishig25 committed:
Update src/transformers/models/layoutlmv2/feature_extraction_layoutlmv2.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

mishig25 committed:
Update src/transformers/models/layoutlmv2/feature_extraction_layoutlmv2.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

mishig25 committed:
Update src/transformers/models/layoutlmv2/feature_extraction_layoutlmv2.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

mishig25 committed:
Chore

mishig25 committed:
Merge branch 'layout_detection' of github.com:mishig25/transformers into layout_detection

mishig25 committed:
Add comments and make fixup

mishig25 commented on src/transformers/pipelines/object_detection.py:
```diff
@@ -108,22 +108,42 @@ def preprocess(self, image):
         model_outputs = {"outputs": outputs, "target_size": target_size}
+        if self._is_layout_detection():
+            model_outputs["bbox"] = model_inputs["bbox"]
+            model_outputs["offset_mapping"] = offset_mapping
```
>  because we're not depending on the model atm
Makes so much sense!
Pushed the suggested changes

mishig25 commented on src/transformers/pipelines/object_detection.py:
```diff
@@ -108,22 +108,42 @@ def preprocess(self, image):
+        post_process_inputs = [model_outputs["outputs"], model_outputs["target_size"]]
+        if self._is_layout_detection():
+            post_process_inputs.extend([model_outputs["offset_mapping"], model_outputs["bbox"]])
+        raw_annotations = self.feature_extractor.post_process(*post_process_inputs)
```
Done

mishig25 committed:
Merge branch 'master' into layout_detection

mishig25 committed:
Refacotr is_subwords

mishig25 commented on src/transformers/models/layoutlmv2/feature_extraction_layoutlmv2.py:
```diff
@@ -220,3 +233,54 @@ def __call__(
+        labels = labels.tolist()
+        boxes = bbox.tolist()
+        target_sizes = target_sizes.tolist()
+        is_subword = np.array(offset_mapping.tolist())[:, :, 0] != 0
```
Did some refactoring and added comments. Wdyt
It works and was used in the test_method below 😊
https://github.com/huggingface/transformers/blob/64c2df5570ff97024d60500570055ff44361f142/src/transformers/models/layoutlmv2/feature_extraction_layoutlmv2.py#L274-L281

Narsil replied:
It's much better. Looking at the actual data, I understood it's linked to some captured text being multi-token, and we're taking only the first token.

Few things:
- `subword` is not an adapted language IMO. subwords don't necessarily exist for random tokenizers so we should stay away from the term `word` if possible. Here you're talking about "first token". `is_first_token`, and `offsets`.
- `offsets` are offsets, so let's keep their name. IMO `offset_start == 0 # in case of multiple tokens we look only at the beginning of the string` seems much easier to understand. (You do have to know what an offset is, but I think it's easier)

Narsil replied:
And maybe let the door open to having more complex behavior. We're dropping a lot of info here. It's perfectly fine in this PR, but future readers/contributors might want to improve this.

mishig25 replied:
changed the variable names as suggested 👍

NielsRogge commented on src/transformers/models/layoutlmv2/feature_extraction_layoutlmv2.py:
```diff
@@ -220,3 +235,54 @@ def __call__(
+        labels = labels.tolist()
+        boxes = bbox.tolist()
+        target_sizes = target_sizes.tolist()
+        subwords = offset_mapping.tolist()
```
I wouldn't call them subwords, but offsets.

NielsRogge commented on src/transformers/models/auto/modeling_auto.py:
```diff
@@ -263,6 +263,8 @@
     [
         # Model for Object Detection mapping
         ("detr", "DetrForObjectDetection"),
+        # layoutlmv2 is able to detect entities (tokens) within an image with an OCR which we can be interpreted as objects
```
```suggestion
        # layoutlmv2 is able to detect entities (tokens) within an image with an OCR which can be interpreted as objects
```

mishig25 committed:
Update src/transformers/models/auto/modeling_auto.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

mishig25 committed:
Add post_process test and better offset naming

mishig25 committed:
Merge branch 'layout_detection' of github.com:mishig25/transformers into layout_detection

mishig25 committed:
Chore

mishig25 committed:
Set Obj Det default threhsold 0.5

mishig25 committed:
make fixup

mishig25 commented on tests/test_pipelines_object_detection.py:
```diff
@@ -251,3 +254,64 @@ def test_threshold(self):
+        doc_img = dataset[0]["image_path"]
+
+        pipe_inputs = object_detector.preprocess(doc_img)
+        self.assertEqual(
+            set(pipe_inputs.keys()),
+            {"input_ids", "token_type_ids", "attention_mask", "offset_mapping", "bbox", "image", "target_size"},
```
Am I correct to assume that you want actual values like in: 
https://github.com/huggingface/transformers/blob/9a8eadfe7114ba2542a8ca627200c2459ac7a71e/tests/test_pipelines_object_detection.py#L135-L141


If so, what should I do for fields like pipe_inputs["image"] that has size torch.Size([1, 3, 224, 224]) ? Should I use smaller image/input or should I trim output as in:  
https://github.com/huggingface/transformers/blob/9a8eadfe7114ba2542a8ca627200c2459ac7a71e/tests/test_pipelines_object_detection.py#L270-L275

Narsil replied:
Hoooo I read a bit fast, and didn't realize it was intermediary steps

- Slow tests are "integration" test (they test from a higher level overview) and aim to make sure we're not breaking any results in the future (hence real values)
- Fast tests are closer to the code and check lower level (I think `preprocess` and `forward` belong to those).

It feels like this test is a bit in between. I don't think it's a blocker so I can approve but IMHO:

- Either this is a low level test that LayoutConfiguration gets the extra keys, and it needs to be handled within `run_pipeline_test` (and it doesn't actually depend on `LayoutConfiguration` but on `ForTokenClassification`). 
- Or this is an integration tests, and we do need to check the real values (image tensors can be ommitted with something like `ANY(torch.Tensor)` with a comment).

Reading the test, I think the intent was to aim for the first one (making sure extra keys are present).

mishig25 replied:
The intent was indeed the first one (making sure extra keys are present). Therefore, extended `run_pipeline_test` to also test the keys when architecture == ForTokenClassification

https://github.com/huggingface/transformers/blob/3825a4175a184f643085d978c5bafd4518f40fe0/tests/test_pipelines_object_detection.py#L109-L113

mishig25 commented on src/transformers/models/layoutlmv2/feature_extraction_layoutlmv2.py:
```diff
@@ -220,3 +233,54 @@ def __call__(
 
         return encoded_inputs
+
+    def post_process(self, outputs, target_sizes, offset_mapping, bbox):
```
added the test

mishig25 commented:
Currently, there are 2 layout models: [layoutlm](https://github.com/huggingface/transformers/tree/master/src/transformers/models/layoutlm) & [layoutlmv2](https://github.com/huggingface/transformers/tree/master/src/transformers/models/layoutlmv2).

Unlike layoutlmv2, layoutlm lacks its feature_extractor (which should be quite similar to layoutlmv2's one) -> therefore, layoutlm (layoutlmfortokenclassification) can;t be used in ObjectDetectionPipeline currently.
I propose to take care of it (add feature_extractor to layoutlm) in a separate PR

wdyt @Narsil @NielsRogge

NielsRogge commented on tests/test_feature_extraction_layoutlmv2.py:
```diff
@@ -159,6 +159,15 @@ def test_call_pytorch(self):
 
         # Test not batched input
         encoded_images = feature_extractor(image_inputs[0], return_tensors="pt").pixel_values
+        print(encoded_images.shape)
+        print(
+            (
+                1,
+                self.feature_extract_tester.num_channels,
+                self.feature_extract_tester.size,
+                self.feature_extract_tester.size,
+            )
+        )
```
These print statements should be removed

mishig25 replied:
Done 😂

mishig25 committed:
Rm print statements

NielsRogge commented:
@mishig25 you can directly use `LayoutLMv2FeatureExtractor` for LayoutLMv1. You should just not use the `image` key it creates.

mishig25 commented:
@NielsRogge sounds great.

Becuase of the `transformers` design, I will copy all of the code from `models/layoutlmv2/feature_extraction_layoutlmv2.py` -> `models/layoutlm/feature_extraction_layoutlm.py`

 one detail I'd like to know is:
1. [LayoutLMForTokenClassification](https://github.com/huggingface/transformers/blob/eae7a96b7d5810ee6723e41f9b316cc51672fbb1/src/transformers/models/layoutlm/modeling_layoutlm.py#L1082-L1087) exists for the Pytorch version of the architecture only (not for Tensorflow version of the model)
2. LayoutLMv2FeatureExtractor has torch dependency as [in here](https://github.com/huggingface/transformers/blob/f3248a67ca8ed8c1075f7bdcb7ab896a5c59f4d0/src/transformers/models/layoutlmv2/feature_extraction_layoutlmv2.py#L30-L33)
3. Because 1 & 2 -> therefore, I should not add any tensorflow related code to `layoutlm/feature_extraction_layoutlm.py`. Is it correct?

NielsRogge commented:
Ok, perhaps it might be better to do that in a separate PR. Because then you also need to add the tokenizer, etc.

Narsil reviewed (approved):
LGTM.

I left a comment again about the last test which purpose is not entirely clear to me but I think it's OK to move forward as-is too.

Narsil reviewed (commented):
LGTM.

I left a comment again about the last test which purpose is not entirely clear to me but I think it's OK to move forward as-is too.

mishig25 commented on tests/test_feature_extraction_layoutlmv2.py:
```diff
@@ -159,6 +159,15 @@ def test_call_pytorch(self):
+                self.feature_extract_tester.size,
+                self.feature_extract_tester.size,
+            )
+        )
```
indeed, removed it in https://github.com/huggingface/transformers/pull/13622/commits/05db577805e550ba66241700b4e6e25b3f181213

NielsRogge commented on src/transformers/models/layoutlmv2/feature_extraction_layoutlmv2.py:
```diff
@@ -220,3 +235,52 @@ def __call__(
+
+        results = []
+
+        for s, l, b, o, (height, width) in zip(scores, labels, boxes, offsets, target_sizes):
+            # only keep start of a particular word (i.e. offset start == 0) (e.g. [San, Fran, ##cis, ##co] -> [San])
```
s, l, b, o aren't that easily readable, could we use something more informative instead?

Also, for [San, Fran, ##cis, ##co] -> [San], we are actually using the predictions for both [San] and [Fran], as they are both tokens that are at the start of a particular word.
```suggestion
        for s, l, b, o, (height, width) in zip(scores, labels, boxes, offsets, target_sizes):
            # only keep start of a particular word (i.e. offset start == 0) (e.g. [San, Fran, ##cis, ##co] -> [San, Fran])
```

NielsRogge reviewed (approved):
LGTM! Only a minor change in the `post_process` method and it's OK.

mishig25 committed:
Update src/transformers/models/layoutlmv2/feature_extraction_layoutlmv2.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

mishig25 committed:
Better variable names in post_process

Narsil commented on src/transformers/models/layoutlmv2/feature_extraction_layoutlmv2.py:
```diff
@@ -276,7 +276,7 @@ def post_process(self, outputs, target_sizes, offset_mapping, bbox):
 
         for s, l, b, o, (height, width) in zip(scores, labels, boxes, offsets, target_sizes):
-            # only keep start of a particular word (i.e. offset start == 0) (e.g. [San, Fran, ##cis, ##co] -> [San])
+            # only keep start of a particular word (i.e. offset start == 0) (e.g. [San, Fran, ##cis, ##co] -> [San, Fran])
```
Are you sure ?
It depends on the boxes boundary of the OCR no ? If OCR outputs [SAN] [Francisco] then yes,
but if it outputs [San Francisco] it feels like the previous commentary was correct.

No ?

Narsil replied:
Offsets != words....

NielsRogge replied:
Yes you're right @Narsil, it depends on the OCR. Good catch.

mishig25 replied:
Reverted the comment to the previous (correct) version 
https://github.com/huggingface/transformers/blob/3825a4175a184f643085d978c5bafd4518f40fe0/src/transformers/models/layoutlmv2/feature_extraction_layoutlmv2.py#L277-L280

mishig25 committed:
Fix wrong comment

mishig25 committed:
Extend run_pipeline_test test

mishig25 committed:
Fix tests

mishig25 committed:
Merge branch 'master' into layout_detection

mishig25 committed:
Add require backends to tests

mishig25 commented:
@Narsil since I've made [this change](https://github.com/huggingface/transformers/pull/13622#discussion_r715404373),
[ci/circleci: run_tests_pipelines_torch ](https://app.circleci.com/pipelines/github/huggingface/transformers/28325/workflows/6bb6f52b-bbc1-4f38-936f-a1cb31490c1c/jobs/278922) is failing for a reason: `requires_backends(self, "detectron2")`

However, I've tried adding `@require_detectron2` decorator both to `class ObjectDetectionPipelineTests` & method `run_pipeline_test` and wasn't able to make the test pass.

What am I missing?

Narsil commented:
@mishig25 It seems it's unrelated tests. Did you rebase ?

It seems like other generic tests are failing because Layout is declaring himself as ForQuestionAnswering etc... and those tests are ran without detectron2...

Feel free to ignore, I'll take a look

mishig25 commented:
@Narsil I did rebase. It should be unrelated but the tests started failing after commit https://github.com/huggingface/transformers/pull/13622/commits/df86d519d89e69ec2b823cbb36bac58e6bfaac18

mishig25 committed:
Update get_config to use smaller resent

mishig25 committed:
Rm dev change

Narsil committed:
Disabling tests for bimodal models on pipelines that do not support it.

mishig25 committed:
Merge branch 'master' into layout_detection

Narsil committed:
Making all tests pass attempt #1.

Attempt #2.

Limit scope of error exception to detectron2

`Lxmert` is special.

Typo.

mishig25 commented:
@LysandreJik could you please check this https://github.com/huggingface/transformers/pull/13622/commits/c93385e0af44b41373100a81ba44628e1bc60379 and https://github.com/huggingface/transformers/pull/13622/commits/defd5747afdc906d7acc86225d4fe41ca8336398 commits where Nicolas is disabling some tests for a reason:
> Basically Layout displays itself as a valid model for many pipelines but it isn't (like other vision models it's fine) I just disabled those tests as they're not supposed to work
(ForQuestionAnswering, ForTextClassification and so on)

LysandreJik commented on tests/test_pipelines_object_detection.py:
```diff
@@ -46,16 +48,22 @@ def open(*args, **kwargs):
     model_mapping = MODEL_FOR_OBJECT_DETECTION_MAPPING
 
+    @require_detectron2
     @require_datasets
+    @require_pytesseract
     def run_pipeline_test(self, model, tokenizer, feature_extractor):
-        object_detector = ObjectDetectionPipeline(model=model, feature_extractor=feature_extractor)
+        object_detector = ObjectDetectionPipeline(
+            model=model, tokenizer=tokenizer, feature_extractor=feature_extractor
+        )
         outputs = object_detector("./tests/fixtures/tests_samples/COCO/000000039769.png", threshold=0.0)
```
Adding `detectron2` here means that previous object detection models (DETR) will not be run. The CircleCI run that has the `detectron2` dependency only runs the LayoutLM-v2 tests.

mishig25 replied:
@Narsil will address this comment and will merge

mishig25 replied:
@LysandreJik would removing `require_detectron2` & `require_pytesseract` decorators from  `run_pipeline_test` work? (since these decorators are being required at a class level)
https://github.com/huggingface/transformers/blob/e6107347b67537f807cb0f2c4321118da5278271/tests/test_pipelines_object_detection.py#L60-L62

https://github.com/huggingface/transformers/blob/e6107347b67537f807cb0f2c4321118da5278271/tests/test_pipelines_object_detection.py#L51-L57

If not, please let me know what would be the best way to handle this problem

LysandreJik reviewed (commented):
This looks good! The tests removed look fine to me.

Only left a comment relative to a test that would not run anymore for DETR, which is an issue

Narsil commented:
@mishig25 Any blockers left for this ?

mishig25 committed:
Rm layout specific decorators from run_pipeline_ts

LysandreJik commented:
Feel free to merge when ready @mishig25
